pkg/virt-launcher/virtwrap/converter/converter.go

```diff
@@ -325,6 +333,50 @@ func SetOptimalIOMode(disk *api.Disk) error {
 	return nil
 }
 
+// SetOptimalBlockIO sets the BlockIO fields to match the backing device
+func SetOptimalBlockIO(disk *api.Disk) error {
+	if disk.BlockIO != nil {
+		return nil
+	}
+
+	var path string
+	if disk.Source.File != "" {
+		path = disk.Source.File
+	} else if disk.Source.Dev != "" {
+		path = disk.Source.Dev
+	} else {
+		return nil
+	}
+
+	f, err := os.OpenFile(path, syscall.O_RDONLY, 0)
+	if err != nil {
+		return nil
+	}
+	logicalSize, err := unix.IoctlGetInt(int(f.Fd()), unix.BLKBSZGET)
+	if err != nil {
+		return nil
+	}
+	physicalSize, err := unix.IoctlGetInt(int(f.Fd()), unix.BLKSSZGET)
+	if err != nil {
+		return nil
+	}
+	defer util.CloseIOAndCheckErr(f, nil)
```

Comment:
This should go right after a successfull file open. Otherwise you will leak the descriptor if one of the ioctl's fails.

Reply:
You are correct here. I had already changed this locally but haven't pushed that yet.  

In the broader scope of this change/feature not only do I now have a grant total of only like 2h of being in contact with Go code, the same is true for the KubeVirt project code. I still need to spend more time on reading into the inner workings of KubeVirt itself as, unlike the other storage related change, this is spread across several files and touches different parts of the system. It's adding new stuff not just fixing something existing.  

That, as well as the question of whether or not it's worth investing more time on this (would this option/feature even be wanted/desired in KubeVirt? I'd like to have it but what about others/upstream?), is also why I put it on Draft status for now. I highly doubt I've touched all the relevant places and I also think I probably need to update the admission logic so it blocks changes/updates to the sizes after creation. While theoretically you could make certain changes to these sizes without breaking things, if things are properly aligned and match on the new sizes, I think it'd be better to prevent it for existing disks.